### PR TITLE
Refactor remote-routing-table service inline with remote state interfaces

### LINKED ISSUE #14067: [Remote Routing Table] Refactor Remote routing table interfaces
### Is your feature request related to a problem? Please describe

- Refactor Remote routing table interfaces to use remote interfaces from PR https://github.com/opensearch-project/OpenSearch/pull/13785
- Remove deadCode - https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/remote/RemoteStorePathStrategy.java#L110
- basePath to be constructed directly in RemoteRutingTableService since its moved to utils as part of refactoring https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java#L560

### Related component

ShardManagement:Routing

### PR DESCRIPTION
**Pull request by Arpit-Bandejiya**
### Description
The PR refactor the current RemoteRoutingTableService to implement the remote table entities.

### Related Issues
Resolves #14067 

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable. --> N/A

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 1]
```diff
@@ -722,7 +707,7 @@ private UploadedMetadataResults writeMetadataInParallel(
         UploadedMetadataResults response = new UploadedMetadataResults();
         results.forEach((name, uploadedMetadata) -> {
             if (uploadedMetadata.getClass().equals(UploadedIndexMetadata.class)
-                && uploadedMetadata.getComponent().contains(InternalRemoteRoutingTableService.INDEX_ROUTING_METADATA_PREFIX)) {
+                && uploadedMetadata.getComponent().contains(INDEX_ROUTING_TABLE_PREFIX)) {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:710]: We should use the same variable while adding the task and checking here. In case delimiter changes or prefix changes, this will break.

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java [hunk 2]
```diff
@@ -67,64 +55,37 @@
...
-
-    public static final String INDEX_ROUTING_PATH_TOKEN = "index-routing";
-    public static final String INDEX_ROUTING_FILE_PREFIX = "index_routing";
-    public static final String INDEX_ROUTING_METADATA_PREFIX = "indexRouting--";
-
     private static final Logger logger = LogManager.getLogger(InternalRemoteRoutingTableService.class);
     private final Settings settings;
     private final Supplier<RepositoriesService> repositoriesService;
+    private final Compressor compressor;
+    private final RemoteWritableEntityStore<IndexRoutingTable, RemoteIndexRoutingTable> remoteWritableEntityStore;
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:62]: nit: rename `remoteWritableEntityStore` to `remoteIndexRoutingTableStore`

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java [hunk 3]
```diff
@@ -152,42 +113,34 @@ public DiffableUtils.MapDiff<String, IndexRoutingTable, Map<String, IndexRouting
     /**
      * Create async action for writing one {@code IndexRoutingTable} to remote store
      * @param clusterState current cluster state
+     * @param clusterUUID current cluster UUID
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:116]: Do we plan on adding clusterUUID to this method param?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 4]
```diff
@@ -9,92 +9,106 @@
...
 
-    public RemoteIndexRoutingTable(IndexRoutingTable indexRoutingTable) {
+    public RemoteIndexRoutingTable(
+        IndexRoutingTable indexRoutingTable,
+        String clusterUUID,
+        Compressor compressor,
+        long term,
+        long version
+    ) {
+        super(clusterUUID, compressor, null);
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:48]: Rather than passing `null` here, might be better to have a constructor with required params and internally super class can handle to keep other params as `null` or `optional` based on usage

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 5]
```diff
@@ -9,92 +9,106 @@
...
-     * @param streamOutput output stream to write
-     * @throws IOException exception thrown on failing to write to stream.
-     */
+    public Index getIndex() {
+        return index;
+    }
+
+    @Override
+    public BlobPathParameters getBlobPathParameters() {
+        return new BlobPathParameters(List.of(indexRoutingTable.getIndex().getUUID()), "");
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:79]: Do we not need filePrefix here?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 6]
```diff
@@ -9,92 +9,106 @@
...
+    @Override
+    public String getType() {
+        return INDEX_ROUTING_TABLE;
+    }
+
+    @Override
+    public String generateBlobFileName() {
+        return String.join(
+            DELIMITER,
+            INDEX_ROUTING_TABLE_PREFIX,
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:91]: File names should be of format `%s_%s `and not `%s-%s`. Lets keep the format consistent.
For different usecases, lets keep variables separate to avoid confusion.

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 7]
```diff
@@ -9,92 +9,106 @@
...
-                "checksum verification failed - expected: 0x"
-                    + Long.toHexString(expectedChecksum)
-                    + ", got: 0x"
-                    + Long.toHexString(readChecksum)
-            );
-        }
+    @Override
+    public InputStream serialize() throws IOException {
+        return INDEX_ROUTING_TABLE_FORMAT.serialize(indexRoutingTable, generateBlobFileName(), getCompressor()).streamInput();
+    }
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:108]: For my understanding, `serialize` here internally calls `toXContent` which IndexRoutingTable does not implement. How does serialize work in this case?

File: server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java [hunk 8]
```diff
@@ -31,8 +28,8 @@
  * @opensearch.internal
  */
 public interface RemoteRoutingTableService extends LifecycleComponent {
-    public static final DiffableUtils.NonDiffableValueSerializer<String, IndexRoutingTable> CUSTOM_ROUTING_TABLE_VALUE_SERIALIZER =
-        new DiffableUtils.NonDiffableValueSerializer<String, IndexRoutingTable>() {
+    DiffableUtils.NonDiffableValueSerializer<String, IndexRoutingTable> CUSTOM_ROUTING_TABLE_VALUE_SERIALIZER =
+        new DiffableUtils.NonDiffableValueSerializer<>() {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java:32]: Why did we remove the data type from serializer?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 9]
```diff
@@ -9,92 +9,111 @@
...
-        }
+    public RemoteIndexRoutingTable(String blobName, String clusterUUID, Compressor compressor) {
+        super(clusterUUID, compressor);
+        this.index = null;
+        this.term = -1;
+        this.version = -1;
+        this.blobName = blobName;
     }
 
     public IndexRoutingTable getIndexRoutingTable() {
```
Comment (soosinha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:71]: Do we need this now ?

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        this.pathHashAlgo = clusterSettings.get(REMOTE_ROUTING_TABLE_PATH_HASH_ALGO_SETTING);
+        clusterSettings.addSettingsUpdateConsumer(REMOTE_ROUTING_TABLE_PATH_TYPE_SETTING, this::setPathTypeSetting);
+        clusterSettings.addSettingsUpdateConsumer(REMOTE_ROUTING_TABLE_PATH_HASH_ALGO_SETTING, this::setPathHashAlgoSetting);
+    }
+
+    @Override
+    public BlobPath getBlobPathForUpload(final AbstractRemoteWritableBlobEntity<IndexRoutingTable> obj) {
+        assert obj.getBlobPathParameters().getPathTokens().size() == 1 : "Unexpected tokens in RemoteRoutingTableObject";
+        BlobPath indexRoutingPath = getBasePath().add(RemoteClusterStateUtils.encodeString(getClusterName()))
+            .add("cluster-state")
```
Comment (soosinha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java:81]: Lets define this as constant as it is used in multiple places now

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java [hunk 11]
```diff
@@ -31,7 +31,7 @@
  */
 public class RemoteClusterStateBlobStore<T, U extends AbstractRemoteWritableBlobEntity<T>> implements RemoteWritableEntityStore<T, U> {
 
-    private final BlobStoreTransferService transferService;
+    protected final BlobStoreTransferService transferService;
```
Comment (soosinha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java:34]: Why changed to protected ?

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java [hunk 12]
```diff
@@ -214,111 +163,21 @@ public List<ClusterMetadataManifest.UploadedIndexMetadata> getAllUploadedIndices
         return new ArrayList<>(allUploadedIndicesRouting.values());
     }
 
-    private void uploadIndex(
```
Comment (soosinha) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:217]: Nice :)

**Review by Bukhtawar — COMMENTED** [review]
i guess apart from the actual file contents, there isn't any other data path/format changes?

**Comment by Arpit-Bandejiya** [comment]
We have an breaking change as part of this PR:

Earlier the structure of the RemoteIndexRoutingTable looks like this:
```
IndexRoutingTableHeader --> CodecUtil Header - index name
Body --> number of shards - shards info
Footer --> checksum
```

Now it will be simplified to this:
```
CodecUtil Header
Body --> IndexRoutingTable
Footer --> CodeUtil Footer
```

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 1 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:716]: Done

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java (hunk 3 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:116]: Fixed the java docs as per the params

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java [hunk 13]
```diff
@@ -88,9 +88,16 @@ public void readAsync(final U entity, final ActionListener<T> listener) {
...
+    }
+
+    public BlobPath getBasePath() {
+        return blobStoreRepository.basePath();
+    }
+
+    public BlobPath getBlobPathForUpload(final AbstractRemoteWritableBlobEntity<T> obj) {
+        BlobPath blobPath = getBasePath().add(RemoteClusterStateUtils.encodeString(getClusterName()))
             .add("cluster-state")
             .add(obj.clusterUUID());
```
[the remark this answers is not in the document]
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java]: Done.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 1 above)
Comment (himshikha) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: While adding the task we are still using different variable

File: server/src/main/java/org/opensearch/common/remote/BlobPathParameters.java [hunk 14]
```diff
@@ -24,6 +24,10 @@ public BlobPathParameters(final List<String> pathTokens, final String filePrefix
         this.filePrefix = filePrefix;
     }
 
+    public BlobPathParameters(final List<String> pathTokens) {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/common/remote/BlobPathParameters.java:27]: Do we need this now?

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 15]
```diff
@@ -897,10 +888,20 @@ public void start() {
...
+        this.remoteRoutingTableService = RemoteRoutingTableServiceFactory.getService(
+            repositoriesService,
+            settings,
+            clusterSettings,
+            threadpool,
+            blobStoreRepository.getCompressor(),
+            blobStoreTransferService,
+            blobStoreRepository,
+            ClusterName.CLUSTER_NAME_SETTING.get(settings).value()
+        );
```
[the remark this answers is not in the document]
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Done.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 16]
```diff
@@ -931,7 +930,9 @@ public void start() {
             namedWriteableRegistry,
             threadpool
         );
+
         remoteClusterStateCleanupManager.start();
+        remoteRoutingTableService.start();
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:935]: These look like unnecessary changes.

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java [hunk 17]
```diff
@@ -347,136 +352,13 @@ public void testGetIndicesRoutingMapDiffIndexDeleted() {
         assertEquals(indexName, diff.getDeletes().get(0));
     }
 
-    public void testGetIndexRoutingAsyncAction() throws IOException {
```
Comment (himshikha) [inline server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:350]: Why are these tests removed?

Comment (Arpit-Bandejiya) [reply server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:350]: Tests are renamed into the appropriate function name.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 16 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:933]: removed few un-necessary changes.

File: server/src/main/java/org/opensearch/common/remote/BlobPathParameters.java (hunk 14 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/common/remote/BlobPathParameters.java]: Removed.

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java [hunk 18]
```diff
@@ -88,18 +90,27 @@ public void readAsync(final U entity, final ActionListener<T> listener) {
...
+    public BlobPath getBasePath() {
+        return blobStoreRepository.basePath();
+    }
+
+    public BlobPath getBlobPathPrefix(String clusterUUID) {
+        return getBasePath().add(RemoteClusterStateUtils.encodeString(getClusterName())).add(CLUSTER_STATE_PATH_TOKEN).add(clusterUUID);
+    }
+
+    public BlobPath getBlobPathForUpload(final AbstractRemoteWritableBlobEntity<T> obj) {
+        BlobPath blobPath = getBlobPathPrefix(obj.clusterUUID());
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java:106]: We are getting basePath from clusterState repo rather than routing table repo. This can give wrong paths.

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 19]
```diff
@@ -9,92 +9,107 @@
...
-        long readChecksum = in.readLong();
-        if (readChecksum != expectedChecksum) {
-            throw new IOException(
-                "checksum verification failed - expected: 0x"
-                    + Long.toHexString(expectedChecksum)
-                    + ", got: 0x"
-                    + Long.toHexString(readChecksum)
+    @Override
+    public String generateBlobFileName() {
+        if (blobFileName == null) {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:87]: In what cases will we have the filename already present if we are calling generate?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 20]
```diff
@@ -9,92 +9,107 @@
...
+            blobFileName = String.join(
+                DELIMITER,
+                getBlobPathParameters().getFilePrefix(),
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:90]: Do we use BlobPathParameter anywhere else? Otherwise this looks redundant since we are creating this object and then getting filePrefix which could have been directly supplied to this function.

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 21]
```diff
@@ -9,92 +9,107 @@
...
-        }
+    public Index getIndex() {
+        return index;
     }
 
-    public IndexRoutingTable getIndexRoutingTable() {
-        return indexRoutingTable;
+    @Override
+    public BlobPathParameters getBlobPathParameters() {
+        return new BlobPathParameters(List.of(indexRoutingTable.getIndex().getUUID()), INDEX_ROUTING_FILE);
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:77]: `getBlobPathParameters` is used at multiple places. Do we need to create new object everytime?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 22]
```diff
@@ -9,92 +9,107 @@
...
 
 /**
  * Remote store object for IndexRoutingTable
  */
-public class RemoteIndexRoutingTable implements Writeable {
+public class RemoteIndexRoutingTable extends AbstractRemoteWritableBlobEntity<IndexRoutingTable> {
 
-    private final IndexRoutingTable indexRoutingTable;
+    public static final String INDEX_ROUTING_TABLE = "index-routing";
+    public static final String INDEX_ROUTING_TABLE_PREFIX = "index-routing--";
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:33]: Is this used anywhere other than tests?

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 23]
```diff
@@ -197,16 +196,18 @@ public RemoteClusterStateService(
         this.remoteStateStats = new RemotePersistenceStats();
         this.namedWriteableRegistry = namedWriteableRegistry;
         this.indexMetadataUploadListeners = indexMetadataUploadListeners;
+        this.clusterService = clusterService;
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:199]: Do we need this as a class variable now that we are creating service here only?

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java [hunk 24]
```diff
@@ -88,18 +90,27 @@ public void readAsync(final U entity, final ActionListener<T> listener) {
...
-            .add(RemoteClusterStateUtils.encodeString(clusterName))
-            .add("cluster-state")
-            .add(obj.clusterUUID());
+    public String getClusterName() {
+        return clusterName;
+    }
+
+    public BlobPath getBasePath() {
+        return blobStoreRepository.basePath();
+    }
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java:99]: Does this need to be `public`?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java [hunk 25]
```diff
@@ -9,92 +9,107 @@
...
-                    indicesRoutingTable.addIndexShard(indexShardRoutingTable);
-                }
-                verifyCheckSum(in);
-                indexRoutingTable = indicesRoutingTable.build();
-            }
-        } catch (EOFException e) {
-            throw new IOException("Indices Routing table is corrupted", e);
-        }
+    public Index getIndex() {
+        return index;
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:72]: Is this used anywhere?

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+    ) {
+        super(blobStoreTransferService, blobStoreRepository, clusterName, threadPool, executor);
+        this.pathType = clusterSettings.get(REMOTE_ROUTING_TABLE_PATH_TYPE_SETTING);
+        this.pathHashAlgo = clusterSettings.get(REMOTE_ROUTING_TABLE_PATH_HASH_ALGO_SETTING);
+        clusterSettings.addSettingsUpdateConsumer(REMOTE_ROUTING_TABLE_PATH_TYPE_SETTING, this::setPathTypeSetting);
+        clusterSettings.addSettingsUpdateConsumer(REMOTE_ROUTING_TABLE_PATH_HASH_ALGO_SETTING, this::setPathHashAlgoSetting);
+    }
+
+    @Override
+    public BlobPath getBlobPathForUpload(final AbstractRemoteWritableBlobEntity<IndexRoutingTable> obj) {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java:77]: Can we add a test for this?

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java (hunk 19 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java:87]: Currently there are no such cases, this is done to avoid mismatch in blobFileName since we are dependent on timestamp unlike other remoteobjects.

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java (hunk 25 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java]: Removed.

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java (hunk 24 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/model/RemoteClusterStateBlobStore.java]: Removed.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 23 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java]: Removed.

File: server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java (hunk 22 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/routingtable/RemoteIndexRoutingTable.java]: Removed from the tests as well.

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java (hunk 26 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java:77]: Done.

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    // For testing only
+    public RemoteStoreEnums.PathType getPathTypeSetting() {
+        return pathType;
+    }
+
+    // For testing only
+    public RemoteStoreEnums.PathHashAlgorithm getPathHashAlgoSetting() {
+        return pathHashAlgo;
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java:106]: These can be protected if they are used only in unit tests

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java [hunk 28]
```diff
@@ -105,25 +106,28 @@ public void setup() {
...
         clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
         clusterService = mock(ClusterService.class);
+        blobStoreTransferService = mock(BlobStoreTransferService.class);
         when(clusterService.getClusterSettings()).thenReturn(clusterSettings);
         blobStoreRepository = mock(BlobStoreRepository.class);
         when(blobStoreRepository.getCompressor()).thenReturn(new DeflateCompressor());
         blobStore = mock(BlobStore.class);
         blobContainer = mock(BlobContainer.class);
-        when(repositoriesService.repository("routing_repository")).thenReturn(blobStoreRepository);
+        when(repositoriesService.repository(anyString())).thenReturn(blobStoreRepository);
```
Comment (himshikha) [inline server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:115]: Why was this change required?

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java [hunk 29]
```diff
@@ -105,25 +106,28 @@ public void setup() {
...
         when(blobStoreRepository.blobStore()).thenReturn(blobStore);
+        when(blobStore.blobContainer(any())).thenReturn(blobContainer);
```
Comment (himshikha) [inline server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:117]: Can we add expected blobPath to this instead of `any`

File: server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java (hunk 27 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/model/RemoteRoutingTableBlobStore.java:106]: Done in latest revision.

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java (hunk 28 above)
Comment (Arpit-Bandejiya) [reply server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:115]: Did it while the tests were broken, reverted it.

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java (hunk 29 above)
Comment (Arpit-Bandejiya) [reply server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:117]: This will be used by multiple paths, hence keeping it `any()`

Comment (himshikha) [reply server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:117]: Are we verifying the blobPath in individual tests then?

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java [hunk 30]
```diff
@@ -362,6 +203,16 @@ protected void doStart() {
...
         blobStoreRepository = (BlobStoreRepository) repository;
+        compressor = blobStoreRepository.getCompressor();
+
+        this.remoteIndexRoutingTableStore = new RemoteRoutingTableBlobStore<>(
+            new BlobStoreTransferService(blobStoreRepository.blobStore(), threadPool),
+            blobStoreRepository,
+            clusterName,
+            threadPool,
+            ThreadPool.Names.REMOTE_STATE_READ,
+            clusterSettings
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:214]: Should we have a remote routing thread pool, in case we want to enable them separately?

Comment (himshikha) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:214]: We haven't yet added the hybrid support, we either enable both or neither of them. Should we add separate pools when we add hybrid mode support?

File: server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java (hunk 29 above)
Comment (Arpit-Bandejiya) [reply server/src/test/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableServiceTests.java:120]: Added it in `testGetAsyncIndexRoutingWriteAction`.

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java (hunk 30 above)
Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:214]: I don' think it is needed at the moment since the remote routing and remote state are clubbed together under Publication. Moreover we can take this up if we see any issue when we do load testing of the feature.

**Review by himshikha — APPROVED** [review]

**Review by soosinha — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java [hunk 31]
```diff
@@ -214,111 +155,21 @@ public List<ClusterMetadataManifest.UploadedIndexMetadata> getAllUploadedIndices
...
-            blobFileName,
-            index,
-            threadPool.executor(ThreadPool.Names.REMOTE_STATE_READ),
-            ActionListener.wrap(
-                response -> latchedActionListener.onResponse(response.getIndexRoutingTable()),
-                latchedActionListener::onFailure
-            )
+        ActionListener<IndexRoutingTable> actionListener = ActionListener.wrap(
+            latchedActionListener::onResponse,
+            latchedActionListener::onFailure
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:167]: nit: do we want to add debug logging

Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/cluster/routing/remote/InternalRemoteRoutingTableService.java:167]: Added the debug logs in the `RemoteClusterStateService` listener.

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 32]
```diff
@@ -1022,10 +1019,10 @@ ClusterState readClusterStateInParallel(
 
         LatchedActionListener<IndexRoutingTable> routingTableLatchedActionListener = new LatchedActionListener<>(
             ActionListener.wrap(response -> {
-                logger.debug("Successfully read cluster state component from remote");
+                logger.debug(() -> new ParameterizedMessage("Successfully read index-routing for index {}", response.getIndex().getName()));
                 readIndexRoutingTableResults.add(response);
             }, ex -> {
-                logger.error("Failed to read cluster state from remote", ex);
+                logger.error(() -> new ParameterizedMessage("Failed to read index-routing from remote"), ex);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1025]: For each of the logger messages can we add more details around which entity failed

Comment (Arpit-Bandejiya) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1025]: We add in while reading writing through the CheckSumWritableBloblStoreFormat here: https://github.com/opensearch-project/OpenSearch/blob/7f0ff145ee3a9725833512967459f0a62a591d2b/server/src/main/java/org/opensearch/repositories/blobstore/ChecksumWritableBlobStoreFormat.java#L88

**Review by Bukhtawar — APPROVED** [review]

**Comment by Arpit-Bandejiya** [comment]
The security check is failing in other merged PR as well: https://github.com/opensearch-project/OpenSearch/pull/14597
